gh-91860: documentation for typing.dataclass_transform (#92768)
python/cpython: Doc/library/typing.rst

   .. versionadded:: 3.11

.. decorator:: dataclass_transform

>    The :data:`~typing.dataclass_transform` annotation may be used to

Reviewer: It's not an _annotation_ per se; and this is also slightly more concise

```suggestion
   :data:`~typing.dataclass_transform` may be used to
```

>    decorate a function that is itself a decorator, a class, or a metaclass.

Reviewer: This feels slightly less confusing if we put "class" and "metaclass" first -- too many uses of the word "decorator" too close together otherwise :)

```suggestion
   decorate a class, a metaclass, or another function that is itself a decorator.
```

   The presence of ``@dataclass_transform()`` tells a static type checker that the

>    decorated function, class, or metaclass performs runtime "magic" that

Reviewer: No need to repeat what we just laid out above

```suggestion
   decorated object performs runtime "magic" that
```

   transforms a class, endowing it with dataclass-like behaviors.

   Example usage with a decorator function::

>       _T = TypeVar("_T")
> 
>       @dataclass_transform()
>       def create_model(cls: type[_T]) -> type[_T]:

Reviewer: I think it looks less alien to new users of `typing` (who this feature probably won't be targeted towards, but still) if we don't use leading underscores for `TypeVar` names

```suggestion
      T = TypeVar("T")

      @dataclass_transform()
      def create_model(cls: type[T]) -> type[T]:
```

          ...
          return cls

      @create_model
      class CustomerModel:
          id: int
          name: str

   On a base class::

      @dataclass_transform()
      class ModelBase: ...

      class CustomerModel(ModelBase):
          id: int
          name: str

   On a metaclass::

      @dataclass_transform()
      class ModelMeta(type): ...

      class ModelBase(metaclass=ModelMeta): ...

      class CustomerModel(ModelBase):
          id: int
          name: str

   Each of the ``CustomerModel`` classes defined in this example will now

>    behave similarly to a dataclass created with the ``@dataclasses.dataclass``
>    decorator. For example, the type checker will synthesize an ``__init__``
>    method.

Reviewer: May as well cross-link to the dataclass docs here. Also, make it clear that this has no behaviour at runtime. Also also: "synthesize" sounds quite type-checker jargony :)

```suggestion
   be treated by a type-checker similarly to a dataclass created with the :func:`@dataclasses.dataclass <dataclasses.dataclass>`
   decorator. For example, the type checker will assume that an ``__init__``
   method will be autogenerated by the runtime.
```

   The arguments to this decorator can be used to customize this behavior:

   * ``eq_default`` indicates whether the ``eq`` parameter is assumed to be

>      True or False if it is omitted by the caller.

Reviewer 2: optionally some backticks around True/False

   * ``order_default`` indicates whether the ``order`` parameter is
     assumed to be True or False if it is omitted by the caller.
   * ``kw_only_default`` indicates whether the ``kw_only`` parameter is
     assumed to be True or False if it is omitted by the caller.
   * ``field_specifiers`` specifies a static list of supported classes
     or functions that describe fields, similar to ``dataclasses.field()``.

> 

Reviewer 3: Also mention that it takes arbitrary other kwargs, as a mechanism to allow future extensions.

.. decorator:: dataclass_transform

   :data:`~typing.dataclass_transform` may be used to
   decorate a class, metaclass, or a function that is itself a decorator.
   The presence of ``@dataclass_transform()`` tells a static type checker that the
   decorated object performs runtime "magic" that
   transforms a class, giving it :func:`dataclasses.dataclass`-like behaviors.

   Example usage with a decorator function::

      T = TypeVar("T")

      @dataclass_transform()
      def create_model(cls: type[T]) -> type[T]:
          ...
          return cls

      @create_model
      class CustomerModel:
          id: int
          name: str

   On a base class::

      @dataclass_transform()
      class ModelBase: ...

      class CustomerModel(ModelBase):
          id: int
          name: str

   On a metaclass::

      @dataclass_transform()
      class ModelMeta(type): ...

      class ModelBase(metaclass=ModelMeta): ...

      class CustomerModel(ModelBase):
          id: int
          name: str

   The ``CustomerModel`` classes defined above will
   be treated by type checkers similarly to classes created with
   :func:`@dataclasses.dataclass <dataclasses.dataclass>`.

>    For example, type checkers will assume these classes have an
>    ``__init__`` method that accepts ``id`` and ``name``.

Reviewer: Nit

```suggestion
   For example, type checkers will assume these classes have
   ``__init__`` methods that accept ``id`` and ``name``.
```
